sips/sip-021/sip-021-trustless-two-way-peg-for-bitcoin.md

```diff
@@ -0,0 +1,1385 @@
+fewer than 6 Stacks blocks _anywhere_. 
+
+* The highest frozen block-commit's
+height monotonically increases.  When a new Stacks block is mined, its parent
+will either be the canonical Stacks tip, or it will be a sibling of an
+unconfirmed Stacks block.  In the first case, the highest frozen block-commit's
+height increases by 1.  In the second case, the highest frozen block-commit's
+height is unchanged.
+
+* When mining off of unblessed block-commits, the highest
+frozen block-commit at Bitcoin height N will either be the same as the one in
+Bitcoin height N+1, or it will be the parent of the one in Bitcoin height N+1.
+This is _not true_ when considering blessed block-commits.
+
+### Rules
+
+By introducing both lifecycle phases and Stacker blessings, this proposal
+creates a new fork-choice rule for Stacks: the canonical fork is always tipped
+by the highest descendant of the highest frozen block-commit's Stacks block,
+even if the block for it is not yet published.
+
+In doing so, the system creates a new miner rule: a miner can mine off of any
+descendant of the highest frozen block-commit (i.e. any unblessed
+block-commit), and it can mine off of any descendant of a blessed block-commit.
+Miners freely and individually choose which member of this set of block-commits
+to mine off of.  A miner cannot mine off of any other block-commits.
+
+### Finalization
+
+Once a block-commit receives 100 Bitcoin confirmations, it can no longer be
+blessed.
+
+When Stackers fulfill a peg-out request, they invalidate all Stacks forks in
+which the peg-out would result in an inconsistent chain state (i.e. someone's
+sBTC balance becomes negative).  These forks will never be treated as canonical
+as long as they contain such inconsistencies.  As such, the act of Stackers
+fulfilling a peg-out request that materializes in Stacks block B is declaring
+that any Stacks fork that descends from B might be canonical (i.e. the
+canonical fork must include B).
+
+Consequently, the act of processing a peg-out is used to keep the chain alive
+even if someone manages to mine 99 block-commits without disclosing their
+blocks.  As long as there exists at least one valid publicly-disclosed block,
+then Stackers can use it to authenticate and process peg-outs and in doing so,
+require that block to be part of the canonical chain.  This enables miners and
+Stackers to prevent a hidden-block miner from halting the chain while spending
+far fewer resources than the attacker.  It also prevents a 51% attacker from
+denying sBTC transfers and peg-outs: if any sBTC transfer materializes in the
+honest fork, and Stackers process at least one peg-out on that fork, then the
+51% attacker's fork is obliged to _also_ process the _same_ sBTC transfers in
+order to keep their fork canonical.
+
+# Related Work
+
+This section is not done yet.  Notes follow.
+
+* PowPeg in RSK
+
+	* Requires HSMs on a majority of Bitcoin miners (brittle;
+closed-membership)
```

Comment:
The Powpeg protocol does not require that a majority of bitcoin miners run PowHSM nodes, or even that the PowHSM operators be miners at all. See existing PowHSM operator list here: https://rootstock.io/powpeg/

Reply:
That actually makes them more brittle and less secure.  It does not matter what attestations a non-miner or a minority of miners make with PowHSM, because the non-miner / minority coalition is not guaranteed to see the true state of the canonical fork with 100% certainty.  Non-miners / minority coalitions only sees the blocks that the majority of miners and next-hop peers _allow them_ to see.  The PowHSM is reduced to security theater.

Reply:
My comment wasn't contesting any sort of security analysis on PowHSMs, more just pointing out that, contrary to what the bulletpoint I quoted says, Powpeg does not require a majority of bitcoin miners to run HSMs. Just want to make sure this is corrected before this section is finalized.

> The PowHSM is reduced to security theater.

The PowHSM is not security theater, like anything it has tradeoffs. Here the tradeoff is: assuming the private keys cannot be extracted from a quorum of HSMs, a quorum of PowHSM operators cannot steal the BTC locked in the Powpeg, but a majority of Rootstock hashpower can. The majority of Rootstock hashpower can steal by mining a block with a valid header but an invalid withdrawal transaction inside that sends all of the BTC to themselves, and burying this block under a sufficient number of confirmations -- currently 4000 Rootstock confirmations are required before a withdrawal is signed by the PowHSM nodes. During the 4000 block confirmation window, the PowHSM operators could notice what's happening and unplug their PowHSM nodes until the valid chain takes over as the most difficult chain again for some sufficiently deep number of confirmations. If the valid chain never overtakes the invalid chain in block depth, then the network could hard fork to get away from the attacking miners, or wait until the [emergency multisig](https://github.com/rsksmart/RSKIPs/blob/master/IPs/RSKIP201.md) kicks in.

Depending on how much of the bitcoin hashrate is mining Rootstock, the Powpeg protocol may be considered strictly superior to a "vanilla" multisig, since bitcoin miners are a highly decentralized group (this will be improved even more with widespread Stratum v2 adoption). Thus the security of the Powpeg approaches that of a hashrate escrow. Whether or not a hashrate escrow _actually is_ more secure than a vanilla multisig can be debated, but at least some people (including the Rootstock community and others such as the drivechain community) feel that way. So to them the Powpeg is a meaningful improvement (not "theater") compared to a vanilla multisig.